### ISSUE TITLE
Allow to customize the mobile layout threshold, or to turn it off in settings

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#9567]
### Problem

A user reported frustration when using Lab with the new feature (#8456) that changes the layout depending on available window width. When they tried to move the browser window to side (split screen) they experienced an unexpected layout change.

### Proposed Solution

The threshold for the layout change is currently hard-coded at `760px`:

https://github.com/jupyterlab/jupyterlab/blob/70509274272d97ce985adf7dd4985a576982322f/packages/application/src/lab.ts#L310

It would be neat to allow to modify it in settings, and maybe also have a switch completely disabling this behavior in the `View` menu.

### Additional context

- JupyterLab 3.0
- Original report: https://stackoverflow.com/questions/65564392/jupyter-lab-keeps-turning-on-simple-mode

**Comment by krassowski** [comment]
Relevant comment by @vidartf from the original PR (https://github.com/jupyterlab/jupyterlab/pull/8456#issuecomment-632647688):

> Nice!  In a future PR, we would probably want to make the format be overridable by the user ("desktop"/"mobile"/"auto" ?). As long as we ensure the API at this point can accommodate that in a non-breaking way I'm happy for this to be the starting point!

**Comment by jasongrout** [comment]
If we don't hide the status bar in the mobile mode, the user can switch off Simple Interface with the switch there.

**Comment by edvolsson** [comment]
> If we don't hide the status bar in the mobile mode, the user can switch off Simple Interface with the switch there.
> 
> _Edit: that would be a quick way to work around the issues people are experiencing, but I still agree we need to have a more in-depth conversation about it_

I am one of the people that finds this feature annoying. I frequently push my window to the side when working, and it drives me insane to have to constantly press that button so I wouldn't even consider that a "quick fix".

**Comment by jasongrout** [comment]
Balancing several concerns here:

1. Nice default for mobile of the Simple Interface
2. Not changing to Simple Interface automatically as you resize the window
3. Respecting the user's explicit setting (as opposed to using the default)

How about this logic?

1. The default for Simple Interface is 'auto', which means on application load (only!), we check the width of the window, and activate it or deactivate it based on the width of the window
2. We *never* change to or from Simple Interface automatically after application load
3. If the user has an explicit setting for Simple Interface, just use that, even on application load.

**Comment by krassowski** [comment]
**Note**: currently resizing down enables the simple mode and hides the status bar but resizing up only shows the status bar but does not disable the simple mode, which may be part of the issue. Was this intended?

![resize_issue](https://user-images.githubusercontent.com/5832902/104057916-cf79df80-51ea-11eb-80f9-bb76f77e382d.gif)

**Comment by jasongrout** [comment]
> **Note**: currently resizing down enables the simple mode and hides the status bar but resizing up only shows the status bar but does not disable the simple mode, which may be part of the issue. Was this intended?

I noticed that too and was very confused by it. That's why I think the automatic behavior should *only* happen at application startup. It's hard to know when expanding if the user had intentionally switched to simple interface, or if we automatically switched to simple interface, since there is only one setting.

**Comment by krassowski** [comment]
> The default for Simple Interface is 'auto', which means on application load (only!), we check the width of the window, and activate it or deactivate it based on the width of the window

This could lead to user confusion as well if their desktop windowing system is configured to open windows as 1/2 of available size of the screen for example. It might work for browsers assuming those maximised most of the time (and generally users can change the size prior to opening lab), but some  JupyterLab users (including myself) may have it configured as a standalone webapp running in a separate window (or use [jupyterlab-app](https://github.com/jupyterlab/jupyterlab_app) or its successor) and then its all out of their control (in control of the windowing system window spawning mechanism).

What about keeping the current behaviour, but:
- disabling the simple mode when resizing up (if not explicitly toggled by user)
- having an override in the settings allowing users to disable the autodection feature altogether, pinning to "mobile" or "desktop" (default "auto")
- have a toggle in the top right part of the webpage saying "Mobile" when in the mobile layout mode (there is a lot of free space there right now) allowing to toggle to "desktop", overriding the mode autodetected from window width?
   - this toggle would not affect the global override, and would not be affected by it allowing to temporarily escape the mobile mode irrespective of the the global override whether its "auto" (default) or "mobile" ("desktop" would not result in it showing up anyways)

**Comment by jasongrout** [comment]
> have a toggle in the top right part of the webpage saying "Mobile" when in the mobile layout mode (there is a lot of free space there right now) allowing to toggle to "desktop", overriding the mode autodetected from window width?

FYI, we did have the Simple Interface toggle in the upper right, and people thought it was better in the status bar.

We are conflating Simple Interface with Mobile right now, but I think it may be better to disentangle the two. For example, GitHub does not assume you are on a mobile device if you resize your screen small, and it provides easy ways to enable/disable mobile phone only if you are on a phone (that I've seen).

Mobile vs Desktop is more than just screen size - there are other things like larger buttons, etc on mobile. So having a true "mobile" mode that is enabled/disabled independently from the Simple Interface, which uses something (browser detection?) to figure out a good default behavior. It may be that mobile mode enables Simple Interface and *also* does other stuff.

For now, I would probably take out the auto-changing altogether, and instead make it detect if you are truly on a "mobile" device.

**Comment by krassowski** [comment]
> FYI, we did have the Simple Interface toggle in the upper right, and people thought it was better in the status bar.

So I was thinking about keeping the "Simple" switch on the status bar where it is, but adding another switch for the "mobile" mode in the upper right (because status bar is hidden in the mobile and to make the distinction).

**Comment by jasongrout** [comment]
> but adding another switch for the "mobile" mode in the upper right (because status bar is hidden in the mobile and to make the distinction).

And that switch in the upper right only appears when actually on mobile?


> Yes; I assumed that currently mobile = simple + no status bar

Presumably it should also make menu items bigger targets, etc.

**Comment by krassowski** [comment]
Another annoyance is that the simple mode makes the `New Launcher` button open new tab/window. It can be counter productive when running JupyterLab as a chrome-app (or within electron) as:
- it will open a new browser window and switch focus;
- this is contrary to my expectation as I already learned that switching the simple mode does not destry the remaining main area widgets that I have open, just hides them:
   - I can switch between the main area widgets with shortcuts (I have them set up to <kbd>Ctrl</kbd> + <kbd>Tab</kbd>/<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>Tab</kbd> to mimic the behaviour of most browsers)
   - I can open new main area widgets (I just do not see the bar with the tabs):
      - Settings → Advanced Settings Editor opens as a new main area widget in current window
      - both files and notebooks from the file browser open as main area widgets in current window

I think that the intent was to make it more like `notebook`, but I do not like this behaviour as default. As a user I would appreciate an option to turn it off easily, but maybe it should not be there at all as it is not consistent with other actions. Or maybe there should be two `New Launcher` buttons when in the simple mode (one opening in the current workspace/window, and another creating a new one); this could also be the same button with a modifier (like "hold shift to open in a new window").

**Comment by jasongrout** [comment]
> Another annoyance is that the simple mode makes the `New Launcher` button open new tab/window. It can be counter productive when running JupyterLab as a chrome-app (or within electron) as:

Before 3.0 was released, we reverted a change that opened new files in a new window to polish it some more. It looks like we missed reverting this behavior too. I'll look into it.

**Comment by jasongrout** [comment]
> Another annoyance is that the simple mode makes the New Launcher button open new tab/window. It can be counter productive when running JupyterLab as a chrome-app (or within electron) as:


Fixed in https://github.com/jupyterlab/jupyterlab/pull/9664 - @krassowski, do you want to review it?

**Comment by AlexMV12** [comment]
Hello,

Thank you for this work and your time.
Just wanted to ask if there are progress on this.

I, similarly to @edvolsson, simply don't want Simple mode to be enabled __ever__. In my workflow I prefer to always have the tab selector, because I work with multiple tabs and I use a tiling window manager. This means that, often, I open a window and automatically the browser gets resized.

When this happens, Simple mode is activated and I have to manually disable that, again and again.

I have not found a way to disable this behaviour, and I'd like to have that.

Thank you,
Alessandro

**Comment by jasongrout** [comment]
I was just looking at this last night again. Looking at the code in https://github.com/jupyterlab/jupyterlab/blob/ae9300a4eaf09fa42f17ca56c7409104721649d6/packages/application/src/lab.ts, I observed:

1. Changes are only made when switching to mobile state, but not when switching to desktop state.
2. The only changes are hiding some elements, closing sidebars, i.e., there is nothing really special about it right now that the user couldn't do themselves in 3-4 steps. In other words, there are no state-based changes like font sizes or mouse target sizes, etc. As such, it's not really a state per se as much as it just tries to be a handy shortcut to doing a few things to make things easier to see online. That kind of explains why it doesn't want to undo the changes when switching to desktop mode, since it won't know if the user has taken these actions separately from the auto switching actions, and we don't want to override an intentional user choice.
2. The detection logic is looking for 760px wide screens. That's actually pretty wide to do an auto-switch, I think.

I would propose that we:

- disable the auto-changing of the mode (I think this is the most important change proposed - the current auto-switching is degrading the desktop experience)
- Possibly make a command and/or menu item that switches the state. This shouldn't be a "toggle" in the menu at this point, since there isn't really anything that is toggled here (i.e., changes are not reverted when switching back to desktop mode). Perhaps the item can be named something like "Configure for mobile", something that conveys that it is going to do a few things for you in the ux that you could do yourself, but it is just a convenience at this point.
- Reconsider and extend the changes made for mobile. For example, do font sizes need to change automatically? Do mouse target sizes need to change since you are usually interacting with a finger instead of a high-precision pointer? Should we change the sidebar to overlap and slide over the main area rather than push the main area over?

**Comment by krassowski** [comment]
Another idea: #9464 uses `data-format='mobile'`, would it be a more appropriate selector rather than a width-based one?

Also there is a bias of users who enjoy the simple mode not reporting issues so maybe we should think how to keep it. Can we still keep them happy? Cross-referencing https://discourse.jupyter.org/t/focus-mode-intern-project/1552/16.

**Comment by jasongrout** [comment]
> Can we still keep them happy?

Note that we aren't taking away *anything* by taking out the auto-switching at this point. The user can still toggle simple interface, can still hide sidebars, hide the status bar, etc. We're just making it a little less convenient since the user would have to toggle each of those items separately and manually.

**Comment by krassowski** [comment]
Yes, I get it, but if the name will be changed to "Configure for mobile" then the reaction of a confused user might be "they removed the simple mode that I really liked". I appreciate that the mobile and simple modes should be split, just wondering how to name things in the UI to make the changes less confusing for users. One approach would be to create two commands "Configure for mobile" and "Switch simple/focus mode" which would initially have large overlap but the would be allowed to diverge. Both would be mutually exclusive.

Another approach would be to instead fix switching back from the simple mode and remove the width-based switching (or even better, make it opt-in) without adding "configure for mobile" command; the mobile UI would be activated solely by `data-format='mobile'`.

**Comment by jasongrout** [comment]
@ellisonbg, @afshin, and I discussed this a bit offline (we happened to all be in an unrelated video conference). The consensus from our discussion was:

In the short term (i.e., patch release to fix the user disruptions):
1. The auto-switching behavior should be deleted
2. The event listeners should be deleted, so even if it switches for other reasons, the status bar, simple interface, etc. don't get changed
3. The application `format` attribute needs to stay to not break backwards compatibility
4. We should step back and take a more holistic approach to responsive layout, simple interface, and mobile support, not conflate the three issues, as a 4.0 roadmap item (@afshin is writing up a 4.0 roadmap item).

Given that, and the discussions here, I'll put in a PR that removes the auto-switching behavior and the listeners for the `format` change, and in that PR I won't be exposing this in the menu, as we may change the api as we reconsider these topics.
